src/dataflow/src/decode/avro.rs

```diff
@@ -55,16 +75,19 @@ impl AvroDecoderState {
 
 impl DecoderState for AvroDecoderState {
     fn decode_key(&mut self, bytes: &[u8]) -> Result<Option<Row>, String> {
-        match block_on(self.decoder.decode(bytes, None, None)) {
-            Ok(Some(row)) => {
-                self.events_success += 1;
-                Ok(Some(row))
-            }
-            Ok(None) => Ok(None),
-            Err(err) => {
-                self.events_error += 1;
-                Err(format!("avro deserialization error: {}", err))
-            }
+        match self.key_decoder.as_mut() {
+            Some(key_decoder) => match block_on(key_decoder.decode(bytes, None, None)) {
+                Ok(Some(row)) => {
+                    self.events_success += 1;
+                    Ok(Some(row))
+                }
+                Ok(None) => Ok(None),
+                Err(err) => {
+                    self.events_error += 1;
+                    Err(format!("avro deserialization error: {}", err))
+                }
+            },
+            None => Err(format!("No key decoder configured")),
```

Comment:
When can this happen? If we expect never, we should panic instead.

Reply:
We can validate this is in purify, and turn this into a panic.

Reply:
Actually this is currently very hard to verify for upsert envelopes, currently, because of its desire to push the key into the value. I can add some stuff to do the verification, but the goal is to delete all the code, and since this isn't less safe right now my inclination is to leave it as is until we can rationalize the upsert key/value stuff.

Reply:
So, when concretely can this happen?

Reply:
Fixing Key/Value transports is the only prerequisite for that. I'm currently working on implementing #6344 , after which it should be straightforward.

Reply:
In terms of wall-clock time: optimistically the PR could be opened next week.

Reply:
Sorry, my use of the word "when" was not clear.

I meant: under what circumstances is it possible to get this "No key decoder configured" message? What sequence of  things can I type into Materialize in order to see that error?

Reply:
Ah, approximately: `CREATE SOURCE FROM KAFKA FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '..'` and then ensure that the confluent schema registry does not have a key schema for your topic.